File: src/growth_curve.py
"""Helpers for synthetic OD600 growth curve examples."""

import numpy as np
import pandas as pd


def _require_columns(df: pd.DataFrame, columns) -> None:
    missing = [col for col in columns if col not in df.columns]
    if missing:
        raise ValueError(f"Missing required columns: {missing}")


def summarize_growth(df: pd.DataFrame) -> pd.DataFrame:
    """Summarize OD600 by strain and time."""
    _require_columns(df, ["strain", "time_hr", "od600"])
    summary = (
        df.groupby(["strain", "time_hr"])
        .agg(mean_od600=("od600", "mean"), sd_od600=("od600", "std"), n=("od600", "count"))
        .reset_index()
    )
    summary["sem_od600"] = summary["sd_od600"] / np.sqrt(summary["n"])
    return summary


def estimate_log_phase_rate(summary_df: pd.DataFrame, start_hr: float = 2, end_hr: float = 6) -> pd.DataFrame:
    """Estimate a simple log-phase rate from ln(OD600) over a chosen time window."""
    _require_columns(summary_df, ["strain", "time_hr", "mean_od600"])
    rows = []
    for strain, group in summary_df.groupby("strain"):
        window = group[(group["time_hr"] >= start_hr) & (group["time_hr"] <= end_hr)]
        if len(window) < 2:
            continue
        slope, intercept = np.polyfit(window["time_hr"], np.log(window["mean_od600"]), 1)
        rows.append({"strain": strain, "log_phase_rate_per_hr": slope, "doubling_time_hr": np.log(2) / slope})
    return pd.DataFrame(rows)

File: R/growth_curve.R
library(tidyverse)

# Helpers for synthetic OD600 growth curve examples.

require_columns <- function(df, columns) {
  missing <- setdiff(columns, names(df))
  if (length(missing) > 0) {
    stop(paste("Missing required columns:", paste(missing, collapse = ", ")))
  }
}

summarize_growth <- function(df) {
  require_columns(df, c("strain", "time_hr", "od600"))
  df %>%
    group_by(strain, time_hr) %>%
    summarise(
      mean_od600 = mean(od600, na.rm = TRUE),
      sd_od600 = sd(od600, na.rm = TRUE),
      n = sum(!is.na(od600)),
      sem_od600 = sd_od600 / sqrt(n),
      .groups = "drop"
    )
}

estimate_log_phase_rate <- function(summary_df, start_hr = 2, end_hr = 6) {
  require_columns(summary_df, c("strain", "time_hr", "mean_od600"))
  summary_df %>%
    filter(time_hr >= start_hr, time_hr <= end_hr, mean_od600 > 0) %>%
    group_by(strain) %>%
    summarise(
      log_phase_rate_per_hr = coef(lm(log(mean_od600) ~ time_hr))[2],
      doubling_time_hr = log(2) / log_phase_rate_per_hr,
      .groups = "drop"
    )
}
